### PR TITLE
Feature/wav2vec2 updates

### PR DESCRIPTION
**Pull request by thorinf**
# What does this PR do?

Additions:
- Fine-tuning of wav2vec2 added.
- Inference of wav2vec2 added.

Fixes/Improvements to Pre-training:
- Gumbel temp uses callback, same as Transformers examples, instead of computing in data-loader. Data-loader just uses the value updated in the callback.
- Tensor reduction! Both data-loader and model have been modified to reduce the tensor size going into GS-VQ and loss to minimise the number of unmasked indexes being included. Unmasked indices do not contribute to pre-training loss and are redundant. This significantly improves memory.
- Evaluation will run and log loss.
- GS-VQ computes both perplexities regardless of train/eval setting. This is how fairseq (original implementation) performs it.
- Parallelise method is modified. Some layers have been added for assignment, and doesn't allow user to pipeline without correct layer numbers.

Fixes # (issue)

## Before submitting
- [N] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [Y] Did you make sure to update the documentation with your changes?
- [N] Did you write any new necessary tests?

### CODE REVIEW

File: examples/speech-pretraining/run_pretraining.py [hunk 1]
```diff
@@ -36,19 +37,25 @@
...
     set_seed,
 )
 from transformers.models.wav2vec2.modeling_wav2vec2 import _compute_mask_indices
 from transformers.trainer_utils import get_last_checkpoint
 from transformers.utils import check_min_version
 from transformers.utils.versions import require_version
 
-
 # Will error if the minimal version of Transformers is not installed. Remove at your own risks.
 check_min_version("4.15.0.dev0")
```
Comment (jimypbr) [inline examples/speech-pretraining/run_pretraining.py:49]: min version and required version have different versions

File: examples/speech-recognition/large_8-ipu_config.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (jimypbr) [inline examples/speech-recognition/large_8-ipu_config.json:1]: is this used? can delete or move to hf hub?

File: examples/speech-recognition/base_4-ipu_config.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (jimypbr) [inline examples/speech-recognition/base_4-ipu_config.json:1]: is this used? can delete or move to hf hub?

**Review by jimypbr — CHANGES_REQUESTED** [review]

File: examples/speech-recognition/large_8-ipu_config.json (hunk 2 above)
Comment (jimypbr) [reply examples/speech-recognition/large_8-ipu_config.json:1]: > Could you put the IPUConfig jsons on the graphcore org on the HF hub please?

Sorry. I remember that you have PRs on the HF Hub to update the configs. I guess you will delete these before merge?

File: examples/speech-pretraining/base_4-ipu_config.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/base_4-ipu_config.json:1]: If this is validated, I think it'll be better to put it on the Hub instead of having it as a file in the examples.

File: examples/speech-pretraining/large_8-ipu_config.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/large_8-ipu_config.json:1]: Same comment as above.

File: examples/speech-pretraining/requirements.txt [hunk 6]
```diff
@@ -5,3 +5,5 @@ torchaudio == 0.10.0+cpu
 librosa
 jiwer
 soundfile
+wandb
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/requirements.txt:8]: Why?

File: examples/speech-pretraining/run_pretraining.py [hunk 7]
```diff
@@ -36,19 +37,25 @@
...
 
-require_version("datasets>=1.18.0", "To fix: pip install -r examples/speech-pretraining/requirements.txt")
-
+require_version(
+    "transformers==4.18.0",
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/run_pretraining.py:52]: Do you really require this?

File: examples/speech-pretraining/run_pretraining.py [hunk 8]
```diff
@@ -275,22 +316,38 @@ def __call__(self, features: List[Dict[str, Union[List[int], torch.Tensor]]]) ->
...
-            self.max_gumbel_temperature * self.gumbel_temperature_decay**self.global_step,
-            self.min_gumbel_temperature,
+
+        batch["mask_time_indices"] = torch.tensor(
+            mask_time_indices, dtype=torch.long, device=device
         )
-        self.model.set_gumbel_temperature(gumbel_temperature)
-        self.global_step += 1
-        batch["gumbel_temperature"] = torch.full([batch_size], gumbel_temperature, dtype=torch.float32)
+        batch["sampled_negative_indices"] = torch.tensor(
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/run_pretraining.py:337]: Cant we make things directly in pytorch instead of using numpy? Dont know if this is relevant, but asking!

File: examples/speech-recognition/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,9 @@
+transformers==4.18.0
+datasets >= 1.18.0
+--find-links https://download.pytorch.org/whl/torch_stable.html
+torchaudio == 0.10.0+cpu
+librosa
+jiwer
+soundfile
+wandb
```
Comment (michaelbenayoun) [inline examples/speech-recognition/requirements.txt:8]: Why?

File: examples/speech-recognition/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,9 @@
+transformers==4.18.0
```
Comment (michaelbenayoun) [inline examples/speech-recognition/requirements.txt:1]: Why?

File: examples/speech-recognition/run_inference_ctc.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+from transformers import (
+    AutoModelForCTC,
+    Wav2Vec2Processor,
+    HfArgumentParser,
+)
+from transformers.utils import check_min_version
+from transformers.utils.versions import require_version
+
+# Will error if the minimal version of Transformers is not installed. Remove at your own risks.
+check_min_version("4.15.0.dev0")
```
Comment (michaelbenayoun) [inline examples/speech-recognition/run_inference_ctc.py:37]: Why? Not matching with the requirements.

File: examples/speech-recognition/run_speech_recognition_ctc.py [hunk 12]
```diff
@@ -0,0 +1,760 @@
...
+    HfArgumentParser,
+    Wav2Vec2Processor,
+    set_seed,
+)
+from transformers.trainer_utils import get_last_checkpoint
+from transformers.utils import check_min_version
+from transformers.utils.versions import require_version
+
+# Will error if the minimal version of Transformers is not installed. Remove at your own risks.
+check_min_version("4.15.0.dev0")
```
Comment (michaelbenayoun) [inline examples/speech-recognition/run_speech_recognition_ctc.py:51]: Why? Not matching with the requirements.

**Review by michaelbenayoun — COMMENTED** [review]
Thanks a lot for this huge contribution!
I left a few comments. 
Also quick question: can you check your code formatter? I think the max length allowed is set to 80 for you, which reformats a lot of the code... you should use the configuration we have:
- `setup.cfg`
- `pyproject.toml`

File: examples/speech-pretraining/README.md [hunk 13]
```diff
@@ -83,7 +83,7 @@ python run_pretraining.py \
 	--dataset_name "librispeech_asr" \
 	--dataset_config_name "clean" \
 	--train_split_name "train.100" \
-	--ipu_config_name Graphcore/wav2vec2-base-ipu \
+	--ipu_config_name "Graphcore/wav2vec2-base-ipu" \
 	--output_dir "./wav2vec2-pretrained-base" \
 	--max_duration_in_seconds 20.0 \
```
Comment (jimypbr) [inline examples/speech-pretraining/README.md:88]: should  be 15.6 seconds for base

**Comment by thorinf** [comment]
Changes made, I'll also push configs to model-hub

**Review by jimypbr — APPROVED** [review]
